src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.cs

```diff
@@ -9,56 +10,384 @@ namespace Microsoft.Maui.Controls.Platform
+		//
+		// A lot can change during the process of pushing/popping a page
+		// i.e. Users might change the root page during an appearing event.
+		// So, instead of just bull dozing through the whole sync we perform one
+		// sync step then recalculate the state of affairs and then perform another
+		// until no more sync operations are left.
+		// Typically it's always a good idea to re-evaluate after any async operation has completed
+		async Task SyncPlatformModalStackAsync()
+		{
+			if (!IsModalReady || syncing)
+				return;
+
+			bool syncAgain = false;
+
+			try
+			{
+				syncing = true;
+
+				int popTo;
+
+				for (popTo = 0; popTo < _platformModalPages.Count && popTo < _modalPages.Count; popTo++)
 				{
-					_previousNavModel = null;
-					if (_navModel == null)
-						_navModel = new NavigationModel();
+					if (_platformModalPages[popTo] != _modalPages[popTo].Page)
+					{
+						break;
+					}
+				}
+
+				// This means the modal stacks are already synced so we don't have to do anything
+				if (_platformModalPages.Count == _modalPages.Count && popTo == _platformModalPages.Count)
+					return;
+
+				// This ensures that appearing has fired on the final page that will be visible after 
+				// the sync has finished
+				CurrentPage?.SendAppearing();
+
+				// Pop platform modal pages until we get to the point where the xplat expectation
+				// matches the platform modals
+				if (_platformModalPages.Count > popTo && IsModalReady)
+				{
+					bool animated = false;
+					if (_modalPages.TryGetValue(CurrentPlatformModalPage, out var request))
+					{
+						_modalPages.Remove(CurrentPlatformModalPage);
+						animated = request.IsAnimated;
+					}
+
+					var page = await PopModalPlatformAsync(animated);
+					page.Parent = null;
+					syncAgain = true;
+				}
+
+				if (!syncAgain)
+				{
+					//push any modals that need to be synced
+					var i = _platformModalPages.Count;
+					if (i < _modalPages.Count && IsModalReady)
+					{
+						var nextRequest = _modalPages[i];
+						var nextPage = nextRequest.Page;
+						bool animated = nextRequest.IsAnimated;
+
+						await PushModalPlatformAsync(nextPage, animated);
+						syncAgain = true;
+					}
+				}
+			}
+			finally
+			{
+				// Code has multiple exit points during the sync operation.
+				// So we're using a try/finally to ensure that syncing always 
+				// gets transitioned to false. If more exit points are added at a later point  
+				// we don't have to always worry about the exit point setting syncing to false.
+				syncing = false;
+
+				// syncAgain is only set after a successful operation so we won't hit a case here
+				// where we hit an infinite loop of syncing.
+				if (syncAgain)
+				{
+					await SyncModalStackWhenPlatformIsReadyAsync().ConfigureAwait(false);
+				}
+			}
+		}
+
+		public async Task<Page?> PopModalAsync(bool animated)
+		{
+			Page modal = _modalPages[_modalPages.Count - 1].Page;
```

Comment:
Can _modalPages.Count be 0 here?

Reply:
Yea, this is how it worked in `Forms` also.

if the user tries to `PopModal` when there is no `Modal` then it'll throw an index out of range exception. 

I'll change it to return a more informative exception message.

Reply:
I added a more informative exception message here.